### PR TITLE
+ Add to_json method to AST::Node

### PR DESCRIPTION
**Pull request by rattrayalex**
This is for consumption from the parser in https://github.com/whitequark/parser/pull/442

### CODE REVIEW

File: lib/ast/node.rb [hunk 1]
```diff
@@ -227,6 +227,21 @@ def inspect(indent=0)
     def to_ast
       self
     end
+    
+    # Produces a machine-readable s-expression.
+    #
+    # @return [Array]
+    def to_json
```
Comment (iliabylich) [inline lib/ast/node.rb]: Shouldn't `to_json` method return JSON?

Comment (whitequark) [inline lib/ast/node.rb]: This doesn't really return JSON though, it returns an array. Needs a better name.

Comment (rattrayalex) [reply lib/ast/node.rb]: Good question. I was mistaken in community convention, looking at https://apidock.com/rails/Hash/to_json (I don't write a ton of ruby).

Comment (rattrayalex) [reply lib/ast/node.rb]: Yes, this needs a better name. Thoughts?

Comment (rattrayalex) [reply lib/ast/node.rb]: I was thinking perhaps `to_a`? Could get rid of the `to_s` since I believe `JSON.generate` will handle symbols just fine.

Comment (whitequark) [reply lib/ast/node.rb]: There is already .to_a, and removing any methods would break API.

Comment (rattrayalex) [reply lib/ast/node.rb]: Ah, it looks like `to_a` is taken (just returns `children`). 

Hmm...

Comment (rattrayalex) [reply lib/ast/node.rb]: I assume you'd rather have the actual `JSON.generate` call take place in `parser`, not here – `ast` should probably not `require 'json'`, I assume. 

Is that correct?

Comment (rattrayalex) [reply lib/ast/node.rb]: Maybe `.to_plain`?

Comment (rattrayalex) [reply lib/ast/node.rb]: Updated

Comment (whitequark) [reply lib/ast/node.rb]: Don't like it.

**Comment by rattrayalex** [comment]
Given that `to_s` and `inspect` both produce strings here, I figured `to_json` should both exist and produce a string. 

I created a new `to_sexp_array` helper which is identical to `to_a` but includes the `type` as a symbol as the first element. I'm all ears for other ideas on the name. 

I also added tests. I am relying on Travis to run them as I haven't set this up locally. 

Thanks for the reviews so far!

**Comment by whitequark** [comment]
Get rid of `to_json` since it's redundant with `JSON.dump` and an implementation is trivial and noninvasive. `to_sexp_array` seems fine.

**Comment by rattrayalex** [comment]
Awesome, thanks! I also renamed to `to_sexp_arr` since that was what I found intuitive when writing tests. Hope that's okay.

**Comment by whitequark** [comment]
No, `to_sexp_array` was fine...

**Comment by rattrayalex** [comment]
K I'll move it back haha

**Comment by rattrayalex** [comment]
Sorry for all the back and forth, thanks so much for your patience!@

**Pull request merged**